src/engine/QueryPlanner.cpp

```diff
@@ -434,6 +407,96 @@ QueryExecutionTree QueryPlanner::createExecutionTree(ParsedQuery& pq) const {
   return *final._qet.get();
 }
 
+bool QueryPlanner::checkUsePatternTrick(
+    ParsedQuery& pq, SparqlTriple& patternTrickTriple) const {
+  bool usePatternTrick = false;
+  // Check if the query has the right number of variables for aliases and group
+  // by.
+  if (pq._groupByVariables.size() == 1 && pq._aliases.size() == 1) {
+    const ParsedQuery::Alias& alias = pq._aliases.back();
+    // Check if the alias is a non distinct count alias
+    if (alias._isAggregate &&
+        alias._function.find("DISTINCT") == std::string::npos &&
+        alias._function.find("distinct") == std::string::npos &&
+        (ad_utility::startsWith(alias._function, "COUNT") ||
+         ad_utility::startsWith(alias._function, "count"))) {
+      // look for a HAS_RELATION_PREDICATE triple
+      for (size_t i = 0; i < pq._rootGraphPattern._whereClauseTriples.size();
+           i++) {
+        const SparqlTriple& t = pq._rootGraphPattern._whereClauseTriples[i];
+        // Check that the triples predicates is the HAS_PREDICATE_PREDICATE.
+        // Also check that the triples object matches the aliases input variable
+        // and the group by variable.
+        if (t._p == HAS_PREDICATE_PREDICATE && alias._inVarName == t._o &&
+            pq._groupByVariables[0] == t._o) {
+          // Assume we will use the pattern trick for now but run several more
+          // checks before actually modifying the query.
+          usePatternTrick = true;
+          // check that all selected variables are outputs of
+          // CountAvailablePredicates
+          for (const std::string& s : pq._selectedVariables) {
+            if (s != t._o && s != alias._outVarName) {
+              usePatternTrick = false;
+              break;
+            }
+          }
+          // Check for triples containing the has predicates triples object.
+          for (size_t j = 0;
+               j < pq._rootGraphPattern._whereClauseTriples.size() &&
+               usePatternTrick;
+               j++) {
+            const SparqlTriple& other =
+                pq._rootGraphPattern._whereClauseTriples[j];
+            if (j != i &&
+                (other._s == t._o || other._p == t._o || other._o == t._o)) {
+              usePatternTrick = false;
+            }
+          }
+          // Check for filters on the has predicate triples subject or object
+          for (const SparqlFilter& filter : pq._rootGraphPattern._filters) {
+            if (filter._lhs == t._o || filter._lhs == t._s ||
+                filter._rhs == t._o || filter._rhs == t._s) {
+              usePatternTrick = false;
```

Comment:
Ok, but in this case we still have a a `ql:has-predicate` in the query, am I assuming correctly that without the `userPatternTrick` flag bein set this will be computed in the classic SPARQL way and will still work or does the query fail?

Reply:
If the usePatternTrick flag is not set the query should still work the normal way.